### PR TITLE
[CSL-2258] health status determined by topology, auxx topology policies

### PR DESCRIPTION
**Pull request by avieth**
The health status was appropriate for relays: if they're at max subscriber capacity, then they're unhealth, and the amazon health status check will reflect that. For other types of nodes, this doesn't fit. An edge node should not be considered unhealthy just because it has 0 subscribers out of 0. 

The auxx topology policies are updated so that MPC messages are actually sent out.

### CODE REVIEW

**Comment by avieth** [comment]
@pva701

**Comment by pva701** [comment]
```
[auxx:DEBUG:ThreadId 4] [2018-02-06 10:02:51.16 UTC] Proposing update...
[auxx:INFO:ThreadId 4] [2018-02-06 10:02:51.16 UTC] Announcing proposal with id fade9c18f59a1c827611f346e9a831222e410b1736b801d2bd34310df57f6a04
[auxx:INFO:ThreadId 4] [2018-02-06 10:02:51.16 UTC] Announcing proposal with id fade9c18f59a1c827611f346e9a831222e410b1736b801d2bd34310df57f6a04
[auxx:DEBUG:ThreadId 4] [2018-02-06 10:02:51.16 UTC] self: message MsgMPC OriginSender not enqueued to any nodes of type NodeCore since no such (relevant) peers listed in Peers {peersRoutes = Routes {_routesCore = [], _routesRelay = [[NodeId 127.0.0.1:3000:0]], _routesEdge = []}, peersClassification = fromList [(NodeId 127.0.0.1:3000:0,NodeRelay)]}
[auxx:DEBUG:ThreadId 4] [2018-02-06 10:02:51.16 UTC] self: inFlight = fromList []
[auxx:DEBUG:ThreadId 4] [2018-02-06 10:02:51.17 UTC] self: message MsgMPC OriginSender enqueued to [NodeId 127.0.0.1:3000:0]
[auxx:DEBUG:ThreadId 89] [2018-02-06 10:02:51.17 UTC] self: sending MsgMPC OriginSender to NodeId 127.0.0.1:3000:0
[auxx:WARNING:ThreadId 93] [2018-02-06 10:02:51.17 UTC] Failed to choose appropriate conversation: [Sending 2b -> Conv 5d: endpoint not reported to be used for sending. Our out specs: HandlerSpecs: [43 -> Conv 60, 4 -> Conv 5, 5 -> Conv 4, 6 -> Conv 7, 37 -> Conv 62, 49 -> Conv 5f, 3d -> Conv 61]]
[auxx:WARNING:ThreadId 89] [2018-02-06 10:02:51.17 UTC] self: sending MsgMPC OriginSender to NodeId 127.0.0.1:3000:0 failed with OutSpecNotReported (fromList [(67,ConvHandler {hsReplyType = 96}),(4,ConvHandler {hsReplyType = 5}),(5,ConvHandler {hsReplyType = 4}),(6,ConvHandler {hsReplyType = 7}),(55,ConvHandler {hsReplyType = 98}),(73,ConvHandler {hsReplyType = 95}),(61,ConvHandler {hsReplyType = 97})]) (43,ConvHandler {hsReplyType = 93}) :: SomeException
[auxx:DEBUG:ThreadId 89] [2018-02-06 10:02:51.18 UTC] self: sent MsgMPC OriginSender to NodeId 127.0.0.1:3000:0
Update proposal submitted along with votes, upId: fade9c18f59a1c827611f346e9a831222e410b1736b801d2bd34310df57f6a04
fade9c18f59a1c827611f346e9a831222e410b1736b801d2bd34310df57f6a04
Command execution finished```

@avieth  there are some problems ^

**Comment by avieth** [comment]
@pva701 Thanks for the snippet. The bit about failing to send to a `NodeCore` is a bad error message, as it's not really an error. I'll make a ticket to remove it.

As for the rest: `InSpecs`/`OutSpecs` are causing trouble... It's not enough to simply have the right listener defined, we also have to come up with a special `Word16` key/value pair in a map. It looks like I was mistaken about the update system `OutSpecs`. I have it set to `mempty` with a comment that `Definition of usWorkers plainly shows the out specs = mempty`. Will fix.

**Comment by pva701** [comment]
@avieth note that update proposal doesn't get into the network due to this exception

**Comment by avieth** [comment]
> @avieth note that update proposal doesn't get into the network due to this exception

Yup, `OutSpecs`/`InSpecs` thing is great at making a node *not* work. I just pushed a fix. Try it out.

**Comment by pva701** [comment]
ok, it works

**Comment by avieth** [comment]
> ok, it works

Let's merge it :)

**Comment by pva701** [comment]
@avieth I would like to review it more carefully tomorrow because, to be honest, I think the same trouble may be with the sending transactions

File: lib/src/Pos/Diffusion/Full/Delegation.hs [hunk 1]
```diff
@@ -38,6 +41,17 @@ delegationRelays
     -> [Relay m]
 delegationRelays logic = [ pskHeavyRelay logic ]
 
+-- | 'OutSpecs' for the tx relays, to keep up with the 'InSpecs'/'OutSpecs'
```
Comment (volhovm) [inline lib/src/Pos/Diffusion/Full/Delegation.hs]: Copy-paste.

**Comment by avieth** [comment]
> I would like to review it more carefully tomorrow because, to be honest, I think the same trouble may be with the sending transactions

Sure. In this patch I added all of the missing relay out specs, including those for transactions.

File: lib/src/Pos/Diffusion/Full/Ssc.hs [hunk 2]
```diff
@@ -61,6 +62,18 @@ sscRelays logic =
     , vssCertRelay (postSscVssCert logic)
     ]
 
+-- | 'OutSpecs' for the tx relays, to keep up with the 'InSpecs'/'OutSpecs'
```
Comment (volhovm) [inline lib/src/Pos/Diffusion/Full/Ssc.hs]: Copy-paste.

**Pull request merged**